src/Parsers/IAST.cpp

```diff
@@ -28,34 +28,86 @@ const char * IAST::hilite_none         = "\033[0m";
+    ASTPtr & delete_list_head_reference = next_to_delete_list_head ? *next_to_delete_list_head : delete_list_head_holder;
+
+    /// Move children into intrusive list
+    for (auto & child : children)
     {
-        deleter->push_back(std::move(children));
-        return;
+        /** If two threads remove ASTPtr concurrently,
+          * it is possible that neither thead will see use_count == 1.
+          * It is ok. Will need one more extra stack frame in this case.
+          */
+        if (child.use_count() != 1)
+            continue;
+
+        ASTPtr child_to_delete;
+        child_to_delete.swap(child);
+
+        if (!delete_list_head_reference)
+        {
+            /// Initialize list first time
+            delete_list_head_reference = std::move(child_to_delete);
+            continue;
+        }
+
+        ASTPtr previous_head = std::move(delete_list_head_reference);
+        delete_list_head_reference = std::move(child_to_delete);
+        delete_list_head_reference->next_to_delete = std::move(previous_head);
     }
 
-    std::list<ASTs> queue;
-    queue.push_back(std::move(children));
-    while (!queue.empty())
+    if (!delete_directly)
+        return;
+
+    while (delete_list_head_reference)
     {
-        for (auto & node : queue.front())
+        auto next_node_to_delete = std::move(delete_list_head_reference->next_to_delete);
+
+        /** Pass list head into child before destruction.
+          * It is important to properly handle cases where subclass has member same as one of its children.
+          *
+          * class ASTSubclass : IAST
+          * {
+          *     ASTPtr first_child; /// Same as first child
+          * }
+          *
+          * In such case we must move children into list only in IAST destructor.
+          * If we try to move child to delete children into list before subclasses desruction,
+          * first child use count will be 2.
+          */
         {
-            /// If two threads remove ASTPtr concurrently,
-            /// it is possible that neither thead will see use_count == 1.
-            /// It is ok. Will need one more extra stack frame in this case.
-            if (node.use_count() == 1)
+            ASTPtr child_to_delete;
+            child_to_delete.swap(delete_list_head_reference);
+            child_to_delete->next_to_delete_list_head = &delete_list_head_reference;
+        }
+
+        /** Here we have 2 cases:
+          * 1. Child during deletion does not move any node into list, then just replace current head with next node to delete.
+          * 2. Child during deletion moved some nodes into list, then iterate list and add next to delete node in list tail.
+          */
+        if (delete_list_head_reference == nullptr)
+        {
+            delete_list_head_reference = std::move(next_node_to_delete);
+        }
+        else
+        {
+            ASTPtr delete_list_tail = delete_list_head_reference;
+
+            while (delete_list_tail)
```

Comment:
A small suggestion how to make this loop a bit more concise and with only one check

```
/// Find the end of the list
while (delete_list_tail->next_to_delete != nullptr)
    delete_list_tail = delete_list_tail->next_to_delete;
/// Append next_node_to_delete to the end
delete_list_tail->next_to_delete = std::move(next_node_to_delete);
```